src/dense.jl

```diff
@@ -39,36 +39,46 @@ struct AutoDiffractor <: AbstractADType end
 mode(::AutoDiffractor) = ForwardOrReverseMode()
 
 """
-    AutoEnzyme{M}
+    AutoEnzyme{M,A}
 
 Struct used to select the [Enzyme.jl](https://github.com/EnzymeAD/Enzyme.jl) backend for automatic differentiation.
 
 Defined by [ADTypes.jl](https://github.com/SciML/ADTypes.jl).
 
 # Constructors
 
-    AutoEnzyme(; mode=nothing)
+    AutoEnzyme(; mode::M=nothing, function_annotation::Type{A}=Nothing)
+
+# Type parameters
+
+  - `A` determines how the function `f` to differentiate is passed to Enzyme. It can be:
+
+      + a subtype of `EnzymeCore.Annotation` (like `EnzymeCore.Const` or `EnzymeCore.Duplicated`) to enforce a given annotation
+      + `Nothing` to simply pass `f` and let Enzyme choose the most appropriate annotation
 
 # Fields
 
-  - `mode::M`: can be either
+  - `mode::M` determines the autodiff mode (forward or reverse). It can be:
 
       + an object subtyping `EnzymeCore.Mode` (like `EnzymeCore.Forward` or `EnzymeCore.Reverse`) if a specific mode is required
       + `nothing` to choose the best mode automatically
 """
-struct AutoEnzyme{M} <: AbstractADType
+struct AutoEnzyme{M, A} <: AbstractADType
     mode::M
 end
 
-function AutoEnzyme(; mode::M = nothing) where {M}
-    return AutoEnzyme{M}(mode)
+function AutoEnzyme(;
+        mode::M = nothing, function_annotation::Type{A} = Nothing) where {M, A}
```

Comment:
let's make this more strict to start with, something like

```
A<:Union{Nothing, EnzymeCore.Annotation}
```

Reply:
I can't do this because it would require pulling in EnzymeCore as a dependency of ADTypes, which we want to keep as light-weight and backend-independent as possible

Reply:
could you make the constructor here just always be the one arg constructor, and move the two arg one (with type checking) to the enzymeext?

Reply:
actually fwiw, this might also apply to mode. Maybe just keep this as a 0-arg constructor, and move the kwarg version to the ext?

Reply:
That would be breaking because users may want to use the kwarg constructor with nothing, which doesn't require having the extension loaded

Reply:
I agree than in most cases people who call AutoEnzyme will have EnzymeCore loaded, but it's not a requirement

Reply:
ah right, darn. We should at least try to add a type check to the new arg tho

Reply:
I don't see how to do it without EnzymeCore?

Reply:
I can do it in DI though, since that's mostly how people will use this

Reply:
yeah I think this is a question of what you can do to put as much of the new arg in an extension